### DISCUSSION TITLE
Dependency injection in .NET and `WithParameter` goodie

### DISCUSSION THREAD

**Discussion by MartyIX** [dotnet/runtime#66071] [Q&A] [answered]
Hi,

I would like to port an app using [Autofac](https://github.com/autofac/Autofac) DI container to use [.NET's dependency container](https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection) because MAUI is using the latter DI library.

One thing that I do not know how to rewrite elegantly[^1] is:

```cs
_ = builder.RegisterType<DatabaseMain>().WithParameter(new TypedParameter(typeof(string), appDataFolder)).InstancePerLifetimeScope();
```

where `DatabaseMain` type is defined in the following way:

```cs
public class DatabaseMain
{
    // ...

    public DatabaseMain(string appDataFolder, TypeA typeA, TypeB typeB, TypeC typeC) 
    {
        // ...
    }
}
```

`TypeA`, `TypeB` and `TypeC` are all registered in DI container but `appDataFolder` should not be added to the DIC because it might be then assigned somewhere where I don't want it. Registration of `string` and integral types in DIC seems like a bad idea to me and it's possible only in some scenarios anyway.

Now I know I can write:

```cs
Services.AddSingleton((IServiceProvider provider) => new DatabaseMain(
    appDataFolder,
    provider.GetRequiredService<TypeA>(),
    provider.GetRequiredService<TypeB>(),
    provider.GetRequiredService<TypeC>()
);
```

but it is more verbose than the Autofac's syntax and when I refactor `DatabaseMain`, I need to fix the DI registration as well.

Obviously, I can create some wrapper type `public record AppDataFolder(string dataFolder)` just to please the .NET's DIC but I would rather not do that. Also I do not really wish to create `AppSettings` class with all these parameters as I want `DatabaseMain` to accept what it needs.

Do you know about another not mentioned solution for this problem?

Thanks!

[^1]: `WithParameter` is explained here: https://autofac.readthedocs.io/en/latest/register/parameters.html?highlight=WithParameter#parameters-with-reflection-components

**Comment by pinkfloydx33** [comment, marked as answer]
You can use [`ActivatorUtilities.CreateInstance`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.dependencyinjection.activatorutilities.createinstance?view=dotnet-plat-ext-6.0#microsoft-extensions-dependencyinjection-activatorutilities-createinstance-1(system-iserviceprovider-system-object())) to pass in additional, non-DI parameters

```csharp 
services.AddSingleton(sp =>
   ActivatorUtilities.CreateInstance<DatabaseMain>(sp, appDataFolder)
);
```

There's a few other methods on ActivatorUtilities that can be used to support such scenarios and create factories that are aware of additional parameters.

**Reply by MartyIX** [reply]
Thank you very much for this. I'll give it a try!
